design/migration-path.md

```diff
@@ -0,0 +1,92 @@
+# Proposal: Migration to Conan 2.0 general guidelines
+
+| **Status**        | **Proposed/Accepted/Deprecated**             |
+|:------------------|:---------------------------------------------|
+| **RFC #**         | ####                                         |
+| **Submitted**     | 2020-11-25                                   |
+| **Dependencies**  | RFC #, #                                     |
+
+---
+
+## Summary
+This document defines some general principles about the future upgrade to Conan 2.0 path that users will need to do.
+In summary:
+- Conan recipes will implement a subset of syntax and features in Conan 1.X that is fully operational and can run in Conan 2.0 without requiring changes in conanfiles.
+- The command line in Conan 2.0 will change, in a non 1.X backwards compatible way.
+- Behavior will change, in a non 1.X backwards compatible way.
+- Configuration files might change, in a non 1.X backwards compatible way, but trying to reduce the changes to a minimum.
+- When Conan 2.0 is released, it will become the active series, and Conan 1.X will only get important bug fixes.
+
+
+## Motivation
+Conan 2.0 is a great opportunity to get rid of some existing legacy, outdated or not optimal defaults, and deprecated practices
+and features. The only way to achieve this is by introducing some breaking behavior. On the other hand, it is necessary to
+define an upgrade path that is as smooth as possible, with a special focus on avoiding the split of the ecosystem like it happened
+with Python 2 and 3.
```

Comment:
I think given the listing of "non 1.X backwards compatible" changes above, this is going to set up exactly the split of the ecosystem that happened with Python 2 and 3.

For small projects it looks easy, but for enormous projects in enterprise applications, it'll delay 2.0 adoption. When you have to adapt tools that are wrapped around Conan, change CI, or have to plan to change everything at once, it just becomes painful enough to not migrate.

I think it's much better to evolve rather than break. Keep old commands working, but supply compatible fixups. Make sure the underlying structures (repositories, cache, recipes) are compatible.

Reply:
> I think it's much better to evolve rather than break. Keep old commands working, but supply compatible fixups. Make sure the underlying structures (repositories, cache, recipes) are compatible.

Of course it is, but we have reached one point that we are struggling to evolve. For example, the amount of complicated logic and code to deal both with revisions and without revisions, is so huge, that evolving things take a lot of time, and what is more important, it is also very risky, and we inevitably break too many things for even small improvements and changes. So this is the reason we are proposing to remove support for not using revisions (see the other proposal), and this will be inevitably breaking.